README
TrueWallet-Api version 0.01
===========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

   use TrueWallet::Api qw/:all/;

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2019 by sctnightcore

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.28.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension TrueWallet::Api.

0.01  Thu Mar 21 00:18:48 2019
	- original version; created by h2xs 1.23 with options
		-XAn TrueWallet::Api



Main.pl
use strict;
use warnings;
use FindBin qw( $RealBin );
use lib "$RealBin/lib";
use Data::Dumper;
use TrueWallet qw(getToken getProfile getBalance getActivity topupTW txDetail logout);

Main();
sub Main {
	# login to TrueWallet
	my $email = '';
	my $password = '';
	my $token = getToken($email, $password);
	if (defined $token) {
		# check Balance 
		my $balance = getBalance($token);
		print "You have balance is $balance\n";
		# check Activity [Start - end] // check all Report id 
		my $activity = getActivity($token,'2018-12-28','2019-01-02'); #start , end [YYYY-MM-DD] 
		foreach my $i ( @$activity ) {
   			my $txdetail = txDetail($token, $i->{report_id});
			print "======================================\n";
			printf "Report ID: %s\n", $i->{report_id};
		    printf "Date: %s\n", $i->{date_time};
		    printf "Topup Mode: %s\n",$txdetail->{service_type};
		    printf "To: %s\n", $txdetail->{ref1};
   		    printf "Amount: %s\n", $i->{amount};
		    print "======================================\n";
		}
		#logout from TrueWallet
		my $logout = logout($token);
	} else {
		print "Cannot get token from truewallet\n";
	}
}



lib/TrueWallet.pm
package TrueWallet;
use strict;
use warnings;
use JSON qw( decode_json encode_json);
use Digest::SHA1 qw(sha1 sha1_hex sha1_base64);
use HTTP::Tiny;
use Data::Dumper;
my $http = HTTP::Tiny->new;

BEGIN {
   our @ISA = qw( Exporter );
   our @EXPORT_OK = qw(getToken getProfile getBalance getActivity topupTW txDetail logout);
   require Exporter;
}

sub getToken {
	my ($user, $pass) = @_;
	my %body = (username => "$user", password => sha1_hex($user.$pass), type => 'email');
	my $response = $http->request( 'POST', 'https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/signin', {
			Host => 'mobile-api-gateway.truemoney.com',
			headers => {
				'content-type' => 'application/json'
			},
			content => encode_json \%body
		}
	);
	if ($response->{success}) {
		my $tokenjson = decode_json($response->{content});
		return $tokenjson->{'data'}->{'accessToken'};		
	} else {
		print "Cannot get Token from Truewallet !\n";
	}
}

sub getProfile {
	my ($token) = @_;
	my $response = $http->request( 'GET', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/profile/$token", {
			Host => 'mobile-api-gateway.truemoney.com'
	});

	if ($response->{success}) {
		my $profilejson = decode_json($response->{content});
		return $profilejson->{'data'};		
	} else {
		print "Cannot get Profile from Truewallet !\n";
	}
}

sub getBalance {
	my ($token) = @_;
	my $response = $http->request( 'GET', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/profile/balance/$token", {
			Host => 'mobile-api-gateway.truemoney.com'
	});

	if ($response->{success}) {
		my $balancejson = decode_json($response->{content});
		return $balancejson->{'data'}->{'currentBalance'};
	} else {
		print "Cannot get Balance from Truewallet !\n";
	}
}
sub getActivity {
	my ($token,$start,$end) = @_;
	my $limit = 25;
	my $response = $http->request( 'GET', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/user-profile-composite/v1/users/transactions/history?start_date=$start&end_date=$end&limit=$limit", {
			Host => 'mobile-api-gateway.truemoney.com',
			headers => {
				'Authorization' => $token
			}	
	});

	if ($response->{success}) {
		my $activityjson = decode_json($response->{content});
		my $activity = $activityjson->{'data'}->{'activities'};
		return $activity;
	} else {
		print "Cannot get Activity from Truewallet !\n";
	}

}

#NEED TO BE TEST 
sub topupTW {
	my ($token, $cashcard) = @_;
	my $time = time;
	my $response = $http->request( 'POST', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/topup/mobile/$time/$token/cashcard/$cashcard", {
			Host => 'mobile-api-gateway.truemoney.com',
			headers => {
				'Authorization' => $token
			}
	});
	if ($response->{success}) {
		my $topuptwjson = decode_json($response->{content});
		return $topuptwjson->{'data'};
	} else {
		print "Cannot use topup in Truewallet !\n";
	}
}

sub txDetail {
	my ($token, $id) = @_;
	my $time = time;
	my $response = $http->request( 'GET', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/user-profile-composite/v1/users/transactions/history/detail/$id", {
			Host => 'mobile-api-gateway.truemoney.com',
			headers => {
				'Authorization' => $token
			}
	});
	if ($response->{success}) {
		my $txdetailjson = decode_json($response->{content});
		return $txdetailjson->{'data'};
	} else {
		print "Cannot Get txdetail from Truewallet !\n";
	}	

}

sub logout {
	my ($token) = @_;
	my $response = $http->request( 'POST', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/signout/$token", {
			Host => 'mobile-api-gateway.truemoney.com'
	});
	if ($response->{success}) {
		my $logoutjson = decode_json($response->{content});
		return $logoutjson->{'data'};		
	} else {
		print "Cannot Logout Truewallet !\n";
	}	
}

1;

lib/TrueWallet/Api.pm
package TrueWallet::Api;

use 5.028001;
use strict;
use warnings;
use JSON qw( decode_json encode_json);
use Digest::SHA1 qw(sha1 sha1_hex sha1_base64);
use HTTP::Tiny;
use Data::Dumper;

require Exporter;
our @ISA = qw(Exporter);
our %EXPORT_TAGS = ( 'all' => [ qw(Login get_Profile get_Balance get_Activity topup_CashCard get_TxDetail Logout) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our $VERSION = '0.01';
our $http = HTTP::Tiny->new;

sub Login {
	my ($user, $pass) = @_;
	my %body = (username => "$user", password => sha1_hex($user.$pass), type => 'email');
	my $response = $http->request( 'POST', 'https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/signin', {
			Host => 'mobile-api-gateway.truemoney.com',
			headers => {
				'content-type' => 'application/json'
			},
			content => encode_json \%body
		}
	);
	if ($response->{success}) {
		my $tokenjson = decode_json($response->{content});
		return $tokenjson->{'data'}->{'accessToken'};		
	} else {
		return 1;
	}
}

sub get_Profile {
	my ($token) = @_;
	my $response = $http->request( 'GET', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/profile/$token", {
			Host => 'mobile-api-gateway.truemoney.com'
	});

	if ($response->{success}) {
		my $profilejson = decode_json($response->{content});
		return $profilejson->{'data'};		
	} else {
		return 1;
	}
}

sub get_Balance {
	my ($token) = @_;
	my $response = $http->request( 'GET', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/profile/balance/$token", {
			Host => 'mobile-api-gateway.truemoney.com'
	});

	if ($response->{success}) {
		my $balancejson = decode_json($response->{content});
		return $balancejson->{'data'}->{'currentBalance'};
	} else {
		return 1;
	}
}
sub get_Activity {
	my ($token,$start,$end) = @_;
	my $limit = 25;
	my $response = $http->request( 'GET', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/user-profile-composite/v1/users/transactions/history?start_date=$start&end_date=$end&limit=$limit", {
			Host => 'mobile-api-gateway.truemoney.com',
			headers => {
				'Authorization' => $token
			}	
	});

	if ($response->{success}) {
		my $activityjson = decode_json($response->{content});
		my $activity = $activityjson->{'data'}->{'activities'};
		return $activity;
	} else {
		return 1;
	}

}

#NEED TO BE TEST 
sub topup_CashCard {
	my ($token, $cashcard) = @_;
	my $time = time;
	my $response = $http->request( 'POST', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/topup/mobile/$time/$token/cashcard/$cashcard", {
			Host => 'mobile-api-gateway.truemoney.com',
			headers => {
				'Authorization' => $token
			}
	});
	if ($response->{success}) {
		my $topuptwjson = decode_json($response->{content});
		return $topuptwjson->{'data'};
	} else {
		return 1;
	}
}

sub get_TxDetail {
	my ($token, $id) = @_;
	my $time = time;
	my $response = $http->request( 'GET', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/user-profile-composite/v1/users/transactions/history/detail/$id", {
			Host => 'mobile-api-gateway.truemoney.com',
			headers => {
				'Authorization' => $token
			}
	});
	if ($response->{success}) {
		my $txdetailjson = decode_json($response->{content});
		return $txdetailjson->{'data'};
	} else {
		return 1;
	}	

}

sub Logout {
	my ($token) = @_;
	my $response = $http->request( 'POST', "https://mobile-api-gateway.truemoney.com/mobile-api-gateway/api/v1/signout/$token", {
			Host => 'mobile-api-gateway.truemoney.com'
	});
	if ($response->{success}) {
		my $logoutjson = decode_json($response->{content});
		return $logoutjson->{'data'};		
	} else {
		return 1;
	}	
}
1;
